### PR TITLE
feat: add HTTP(S) proxy support

### PR DESCRIPTION
**Pull request by b3nk3**
This PR proposes the implementation of a http/https proxy support based on [this PR](https://github.com/semantic-release/gitlab/pull/229)

As pointed out by @alexjfischer on the above PR some extra changes had to be made around gitlab url props, which is included.

This is not my original work, rather a rebase and fixup of the original PR.

I would also like to point out, that we have been running this solution for over 6 months in our production environment.

@travi has raised valid points on the other PR about not taking on tech debt, and recommended the same approach as the GitHub plugin, however based on the extra dependencies and therefore peer dependencies, this implementation is cleaner and poses less risk in support and future tech debt. [More info](https://github.com/semantic-release/gitlab/pull/229#issuecomment-993592863)

Thanks in advance for reviewing. Do let me know if I can be of assistance to get this merged in a timely manner.

### CODE REVIEW

**Comment by chessser** [comment]
Would love to have this feature implemented, exactly what my company requires

**Comment by benbenbenbenbenben** [comment]
I'd like this. Make it happe 🚀

**Comment by fgreinacher** [comment]
@travi Reading through the comments in the related MRs and issue I think this is the most promising way forward. 

I don't see a lot of value in aligning with the GItHub plugin because the code paths we touch here are GitLab-specific anyway and would not be part of the common base.

What do you think?

**Comment by travi** [comment]
thanks for bringing my attention back to this one, @fgreinacher, and apologies to everyone for the additional delays. [this comment](https://github.com/semantic-release/gitlab/pull/229#issuecomment-1005065870) brought be around to agreeing that divergence from the github plugin is justified in this case, but i failed to come back to confirm my thoughts or see things through. thank you @naffers for the thorough investigation to raise visibility into the details that made this more clear.

@fgreinacher if you could help move this effort forward, i agree that this path makes sense at this point.

**Comment by fgreinacher** [comment]
@b3nk3 Are you still able to work on this? If so, could you please rebase on latest master so that I can review?

**Comment by b3nk3** [comment]
Hey @fgreinacher, I'll take a look tomorrow.

**Comment by b3nk3** [comment]
> @b3nk3 Are you still able to work on this? If so, could you please rebase on latest master so that I can review?

Done

File: lib/resolve-config.js [hunk 1]
```diff
@@ -1,10 +1,11 @@
-const {castArray, isNil} = require('lodash');
-const urlJoin = require('url-join');
+const { castArray, isNil } = require("lodash");
```
Comment (fgreinacher) [inline lib/resolve-config.js:1]: @b3nk3 Thanks a lot for the quick rebase! Could you please undo all the style changes here?

You can check via `npm run lint` and fix most things via `npm run lint -- --fix`

File: test/resolve-config.test.js [hunk 2]
```diff
@@ -41,102 +56,158 @@ test('Returns user config via environment variables', t => {
...
+        HTTP_PROXY: proxyUrl,
+      },
+    }
+  );
+
+  t.assert(result.proxy.agent.http instanceof HttpProxyAgent);
+  t.assert(result.proxy.agent.http.proxy.origin === proxyUrl);
+});
+
+test("Returns user config via alternative environment variables with https proxy", (t) => {
```
Comment (fgreinacher) [inline test/resolve-config.test.js:109]: Can we add similar test cases for situations where we expect no or a default agent?

- HTTPS GitLab URL and HTTP_PROXY set
- HTTP GitLab URL and HTTPS_PROXY set
- Neither HTTP_PROXY nor HTTPS_PROXY set

File: README.md [hunk 3]
```diff
@@ -56,6 +56,22 @@ Create a [personal access token](https://docs.gitlab.com/ce/user/profile/persona
 | `GL_TOKEN` or `GITLAB_TOKEN`   | **Required.** The token used to authenticate with GitLab. |
 | `GL_URL` or `GITLAB_URL`       | The GitLab endpoint.                                      |
 | `GL_PREFIX` or `GITLAB_PREFIX` | The GitLab API prefix.                                    |
+| `HTTP_PROXY` or `HTTPS_PROXY`  | HTTP or HTTPS proxy to use.                               |
+
+#### Proxy configuration
+The plugin uses [hpagent](https://github.com/delvedor/hpagent#usage) to pass through a proxy. It supports both HTTP and HTTPS proxy.
```
Comment (fgreinacher) [inline README.md:62]: I'd not mentioned the HTTP/HTTPS distinction here:

```suggestion

The plugin uses [hpagent](https://github.com/delvedor/hpagent#usage) to pass requests through a proxy server. 
```

File: README.md [hunk 4]
```diff
@@ -56,6 +56,22 @@ Create a [personal access token](https://docs.gitlab.com/ce/user/profile/persona
...
+```shell script
+HTTP_PROXY=http://proxyurl.com:8080
+HTTPS_PROXY=http://proxyurl.com:8080
+```
+
+hpagent package supports basic authentication as well by embedding the username and password in the proxy URL:
+```shell script
+HTTP_PROXY=http://user:pwd@proxyurl.com:8080
+HTTPS_PROXY=http://user:pwd@proxyurl.com:8080
+```
```
Comment (fgreinacher) [inline README.md:74]: Maybe we can slim this down a bit:

```suggestion
You can configure a proxy server via the `HTTPS_PROXY` environment variable:

```shell script
HTTPS_PROXY=http://proxyurl.com:8080
```

If your proxy server requires authentication you can embed the username and password in the URL:

```shell script
HTTPS_PROXY=http://user:pwd@proxyurl.com:8080
```

If your GitLab instance is exposed via plain HTTP (not recommended!) use `HTTP_PROXY` instead.
```

**Review by fgreinacher — CHANGES_REQUESTED** [review]
Thanks @b3nk3! I left some comments

File: lib/resolve-config.js (hunk 1 above)
Comment (b3nk3) [reply lib/resolve-config.js:1]: Sorry, I had user level Prettier config and failed to catch this during rebase. I'll fix it.

Comment (b3nk3) [reply lib/resolve-config.js]: Done.

File: test/resolve-config.test.js (hunk 2 above)
Comment (b3nk3) [reply test/resolve-config.test.js]: Done.

- Neither HTTP_PROXY nor HTTPS_PROXY set
The above is covered by the test on Line 6 by not passing a proxy config or passing in an empty one. 

Both cases return an empty object. Same when gitlab and proxy urls mismatch.

File: test/resolve-config.test.js [hunk 5]
```diff
@@ -127,6 +255,7 @@ test('Returns user config over GitLab CI/CD environment variables', t => {
   const gitlabUrl = 'https://host.com';
   const gitlabApiPathPrefix = '/api/prefix';
   const assets = ['file.js'];
+  // Const proxy = {};
```
Comment (fgreinacher) [inline test/resolve-config.test.js:258]:
```suggestion
```

File: test/resolve-config.test.js [hunk 6]
```diff
@@ -154,6 +284,7 @@ test('Returns user config via environment variables over GitLab CI/CD environmen
   const gitlabToken = 'TOKEN';
   const gitlabUrl = 'https://host.com';
   const gitlabApiPathPrefix = '/api/prefix';
+  // Const proxy = {};
```
Comment (fgreinacher) [inline test/resolve-config.test.js:287]:
```suggestion
```

File: test/resolve-config.test.js [hunk 7]
```diff
@@ -188,6 +320,7 @@ test('Returns user config via alternative environment variables over GitLab CI/C
...
```
Comment (fgreinacher) [inline test/resolve-config.test.js:323]:
```suggestion
```

File: test/resolve-config.test.js [hunk 8]
```diff
@@ -214,12 +347,14 @@ test('Returns user config via alternative environment variables over GitLab CI/C
...
       milestones: undefined,
       successComment: undefined,
+      proxy: {},
     }
   );
 });
 
 test('Ignore GitLab CI/CD environment variables if not running on GitLab CI/CD', t => {
   const gitlabToken = 'TOKEN';
+  // Const proxy = {};
```
Comment (fgreinacher) [inline test/resolve-config.test.js:357]:
```suggestion
```

File: README.md [hunk 9]
```diff
@@ -56,6 +56,17 @@ Create a [personal access token](https://docs.gitlab.com/ce/user/profile/persona
 | `GL_TOKEN` or `GITLAB_TOKEN`   | **Required.** The token used to authenticate with GitLab. |
 | `GL_URL` or `GITLAB_URL`       | The GitLab endpoint.                                      |
 | `GL_PREFIX` or `GITLAB_PREFIX` | The GitLab API prefix.                                    |
+| `HTTP_PROXY` or `HTTPS_PROXY`  | HTTP or HTTPS proxy to use.                               |
+
+#### Proxy configuration
+
+The plugin uses [hpagent](https://github.com/delvedor/hpagent#usage) to pass requests through a proxy server.
```
Comment (fgreinacher) [inline README.md:63]:
```suggestion
The plugin supports passing requests through a proxy server. 
```

**Review by fgreinacher — APPROVED** [review]

**Pull request merged**